test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz_persistent.cc

```diff
@@ -0,0 +1,137 @@
+#include "test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ExternalProcessing {
+
+namespace {
+// Reset and recreate Envoy and test server after every 2000 fuzzer execs.
+bool fuzzCreateEnvoy(const uint32_t exec) {
+  const uint32_t exec_before_reset_envoy = 2000;
+  return (exec % exec_before_reset_envoy == 0);
+}
+} // namespace
+
+using envoy::extensions::filters::http::ext_proc::v3::ProcessingMode;
+using envoy::service::ext_proc::v3::ProcessingRequest;
+using envoy::service::ext_proc::v3::ProcessingResponse;
+
+DEFINE_PROTO_FUZZER(const test::extensions::filters::http::ext_proc::ExtProcGrpcTestCase& input) {
+  try {
+    TestUtility::validate(input);
+  } catch (const ProtoValidationException& e) {
+    ENVOY_LOG_MISC(debug, "ProtoValidationException: {}", e.what());
+    return;
+  }
+
+  // have separate data providers.
+  FuzzedDataProvider downstream_provider(
+      reinterpret_cast<const uint8_t*>(input.downstream_data().data()),
+      input.downstream_data().size());
+  FuzzedDataProvider ext_proc_provider(
+      reinterpret_cast<const uint8_t*>(input.ext_proc_data().data()), input.ext_proc_data().size());
+
+  // Using persistent Envoy and ext_proc test server.
+  static std::unique_ptr<ExtProcIntegrationFuzz> fuzzer = nullptr;
+  static std::unique_ptr<ExtProcFuzzHelper> fuzz_helper = nullptr;
+  // Protects fuzz_helper which will be accessed by multiple threads.
+  static Thread::MutexBasicLockable fuzz_helper_lock_;
+
+  static uint32_t fuzz_exec_count = 0;
+  // Initialize fuzzer once with IP and gRPC version from environment
+  if (fuzzCreateEnvoy(fuzz_exec_count)) {
+    fuzzer = std::make_unique<ExtProcIntegrationFuzz>(
+        TestEnvironment::getIpVersionsForTest()[0], TestEnvironment::getsGrpcVersionsForTest()[0]);
+  }
+  // Initialize fuzz_helper during every execution.
+  // This will be accessed by the test server which is initialized once.
+  fuzz_helper = std::make_unique<ExtProcFuzzHelper>(&ext_proc_provider);
+
+  // Initialize test server.
+  if (fuzzCreateEnvoy(fuzz_exec_count)) {
+    // This starts an external processor in a separate thread. This allows for the
+    // external process to consume messages in a loop without blocking the fuzz
+    // target from receiving the response.
+    fuzzer->test_processor_.start(
+        fuzzer->ip_version_,
+        [](grpc::ServerReaderWriter<ProcessingResponse, ProcessingRequest>* stream) {
+          while (true) {
+            ProcessingRequest req;
+            if (!stream->Read(&req)) {
+              return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "expected message");
+            }
+
+            fuzz_helper_lock_.lock();
```

Comment:
This could probably be changed to `MutexLock lock(&fuzz_helper_lock_);` to ensure that will allow auto-release of the lock when it is getting out of scope.

Reply:
Thanks for the comments.

This is a static local variable with lifetime as the one fuzz execution.  At same time, the test server while loop will repeatedly run during that exec. If leave it locked until getting out of scope, which is the end of the fuzz execution, it does not work.